### PR TITLE
ath79: add support for ZTE MF282

### PR DESCRIPTION
**Pull request by andyboeh**
The ZTE MF282 is a LTE router used (exclusively?) by the network operator "3".

Specifications
==============

SoC: QCA9563 (775MHz)
RAM: 128MiB
Flash: 8MiB SPI-NOR + 128MiB SPI-NAND
LAN: 1x GBit LAN
LTE: ZTE MF270 (Cat4), detected as P685M
WiFi: QCA9880ac + QCA9560bgn

MAC addresses
=============

LAN: from config
WiFi 1: from config
WiFi 2: +1

Installation
============

TFTP installation using UART is preferred. Disassemble the device and connect serial. Put the initramfs image as openwrt.bin to your TFTP server and configure a static IP of 192.168.1.100. Load the initramfs image by typing:

  setenv serverip 192.168.1.100
  setenv ipaddr 192.168.1.1
  tftpboot 0x82000000 openwrt.bin
  bootm 0x82000000

From this intiramfs boot you can take a backup of the currently installed partitions as no vendor firmware is available for download.

Once booted, transfer the sysupgrade image and run sysupgrade.

LTE Modem
=========

The LTE modem is probably the same as in the MF283+, all instructions apply.

Configuring the connection using modemmanager works properly, the modem provides three serial ports and a QMI CDC ethernet interface.

### CODE REVIEW

**Comment by Leo-PL** [comment]
Looks fine at the first glance. I'm not sure how the maintainers would look at the `/delete-node/` and `/delete-property` statements - they will certainly work, but I think that extracting a common `qca9563_zte_mf28x.dtsi` and keeping the specifics for MF286 series in qca9563_zte_mf286.dtsi including the previous one would look a bit cleaner, IMO.

**Comment by andyboeh** [comment]
There are several other devices with `/delete-node/` and `/delete-property/` statements, e.g. the FRITZ!7530 .dts. For the ath79 target, I see this in `qca9531_glinet_gl-ar300m16.dts` and `qca9531_glinet_gl-ar300m-lite.dts`.

This is certainly something to discuss, but there is then a bit more work to do since there is a lot of repetition in the 281 dts as well. I didn't want to touch all this as I only have the 282 for testing.

**Comment by Leo-PL** [comment]
Okay, understood. I have MF286 and MF286A for testing, so I could check that for you, if you'd like.
Either way:
Reviewed-by: Lech Perczak <lech.perczak@gmail.com>

**Review by Leo-PL — APPROVED** [review]

**Comment by hauke** [comment]
I would prefer if the delete-node is not needed.
Maybe you could remove the entries from the `qca9563_zte_mf286.dtsi` file and move them to the two dts files including this or create a new sub dtsi file.

**Comment by andyboeh** [comment]
OK, let's do it properly then: I will refactor all ZTE MF28x .dts files in order to reduce repetition.

**Comment by andyboeh** [comment]
@hauke Done, but please keep in mind that I only compile-tested the changes to the MF281 and MF286 dts files. This now affects all ZTE MF28x devices.

**Comment by Leo-PL** [comment]
Please give me some time to build and test on MF286 and MF286A, then.

**Comment by Leo-PL** [comment]
I compared compiled device trees before and after this patch. It seems, that MF281 is now missing `/leds/led-0/color` and `/leds/led0/label` properties. They should be restored - values are the same as in MF286. Others - for MF286* are unchanged, so aside the change needed for MF281, +1 from my side.

File: target/linux/ath79/dts/qca9563_zte_mf282.dts
Comment (Leo-PL) [inline target/linux/ath79/dts/qca9563_zte_mf282.dts:30]: This needs adding to MF281 device tree, with blue changed to green.

Comment (andyboeh) [reply target/linux/ath79/dts/qca9563_zte_mf282.dts:30]: Thanks, fixed.

**Comment by andyboeh** [comment]
> I compared compiled device trees before and after this patch. It seems, that MF281 is now missing `/leds/led-0/color` and `/leds/led-0/label` properties. They should be restored - values are the same as in MF286. Others - for MF286* are unchanged, so aside the change needed for MF281, +1 from my side.

Thanks for the review, it's easy to miss this if you don't look closely.

File: target/linux/ath79/dts/qca9563_zte_mf281.dts
Comment (Leo-PL) [inline target/linux/ath79/dts/qca9563_zte_mf281.dts:18]: You forgot to change colour to green here, as it was set before refactoring.

**Comment by Leo-PL** [comment]
Ok, now it's looking fine,
+1.

**Comment by Leo-PL** [comment]
@andyboeh I have a suggestion - you can now configure QMI interface from `ucidef` as well, which means working WAN on the modem on first boot, for example:
```
ucidef_set_interface "wan" device "/dev/cdc-wdm0" protocol "qmi"
```
Works for me with MF28D, I'm preparing changes for MF286 series and MF283+ now.

**Comment by andyboeh** [comment]
Thanks for the hint. I had this already in, but I couldn't get the modem to work with uqmi (it connects, but no data is transferred). With ModemManager, everything is fine so I decided against this.

**Comment by Leo-PL** [comment]
That's strange, for me this worked out of box in MF283+, it's the same modem indeed. ModemManager does some kind of magic with profiles, probably, so you don't have to configure matching APN with `AT+CGDCONT` command.
I would try to configure matching profile with `AT+CGDCONT` and retry with `uqmi`. Bonus: quicker network and LTE attach if there is a proper default bearer configured.

**Comment by hauke** [comment]
Could you split this into two commits please, one doing the refactoring and one adding the new device. 
Do you plan to add the `ucidef_set_interface` entry for the wwan?

**Comment by andyboeh** [comment]
Done, split it into two commits and added the `ucidef_set_interface` entry. This works, i.e. the interface is correctly set up and the device gets an IP address.

**Pull request merged**
